File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
import "slices"
type qnode struct{
    root *TreeNode
    level int
}
func levelOrder(root *TreeNode) [][]int {
    q := make([]qnode, 0)
    ans := make([][]int, 0)    
    
    if root == nil{
      return ans  
    }
    q = append(q, qnode{root, 0})
    for len(q) != 0{
        v := q[len(q)-1]
        q = slices.Delete(q, len(q)-1, len(q))

        if len(ans) == v.level{
            ans = append(ans, []int{})
        }

        ans[v.level] = append(ans[v.level], v.root.Val)
        // push children to que
        if v.root.Right != nil{
            q = append(q, qnode{v.root.Right, v.level + 1})
        }
        if v.root.Left != nil{
            q = append(q, qnode{v.root.Left, v.level + 1})
        }
    }
    return ans
    
    
    
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool check(TreeNode* p, TreeNode* q)
    {
        if((p == nullptr && q != nullptr) || (p != nullptr && q == nullptr) || ((p&&q) && (p->val != q->val)))
        {
            
            return false;
        }        
        if((!p&&!q))
        {
            return true;
        }       
        bool left = check(p->left,q->left);
        bool right = check(p->right,q->right);
        if(left==false || right==false)
        {
            return false;
        }
        return true;
    }
    
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if(!subRoot)
        {
            return true;
        }

        if(!root)
        {
            return false;
        }

        bool cur = check(root,subRoot);
        bool left = isSubtree(root->left,subRoot);
        bool right = isSubtree(root->right,subRoot);


        if(cur||left||right)
        {
            return true;
        }
        return false;
        
        
        
        
    }
};
